### PR TITLE
Add tags to the identity v3 project

### PR DESCRIPTION
**Pull request by randomswdev**
For #1687

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:
https://github.com/openstack/keystone/blob/fe39838f712880c336e18eadf320e7c9e2007448/keystone/api/projects.py#L111
https://github.com/openstack/keystone/blob/fe39838f712880c336e18eadf320e7c9e2007448/keystone/api/projects.py#L233
https://github.com/openstack/keystone/blob/fe39838f712880c336e18eadf320e7c9e2007448/keystone/api/projects.py#L265
https://github.com/openstack/keystone/blob/fe39838f712880c336e18eadf320e7c9e2007448/keystone/api/projects.py#L490

This pull request adds supports for assocaiting tags to projects using the idenetity v3 APIs. The tags are optional thua llowing backward compatibility with OpenStack versions older than Queens: it is enough to not specify tags when creating or updating a project.

### CODE REVIEW

**Comment by randomswdev** [comment]
I looked at the logs of the gophercloud-acceptance-test but I was unable to understand why the test failed. Is there anybody who can help me in locating the failure reasons?

**Comment by jtopjian** [comment]
@randomswdev It's an issue at OpenLab - you can safely ignore

**Comment by kayrus** [comment]
@randomswdev could you please implement tags support for `ListOpts` as well?

**Comment by randomswdev** [comment]
> @randomswdev could you please implement tags support for `ListOpts` as well?

Looking at the (rest api documentation)[https://docs.openstack.org/api-ref/identity/v3/?expanded=list-projects-detail#list-projects], my understanding is that the project list api does not support filtering on tags. @kayrus Can you double check and let me know? In case I’ll add the tags to the ListOpts tomorrow morning.

**Comment by kayrus** [comment]
@randomswdev I think this is it if I'm not mistaken https://github.com/openstack/keystone/blob/fe39838f712880c336e18eadf320e7c9e2007448/keystone/api/projects.py#L130

**Comment by jtopjian** [comment]
I'm not a stickler for full API support. If it's convenient for you, feel free to add tag filtering support for listing. Otherwise, it can always be done in another PR.

**Comment by randomswdev** [comment]
> @randomswdev I think this is it if I'm not mistaken https://github.com/openstack/keystone/blob/fe39838f712880c336e18eadf320e7c9e2007448/keystone/api/projects.py#L130

Added the suggested filters. Sorry for missing them, but the REST API documentation didn't mention them.

**Comment by jtopjian** [comment]
> Sorry for missing them

There's no apology needed. Like I said, it wasn't a requirement to have these included.

Thank you for your work on this. Let me know if this is ready for a code review.

File: openstack/identity/v3/projects/requests.go [hunk 1]
```diff
@@ -153,6 +168,9 @@ type UpdateOpts struct {
 
 	// Description is the description of the project.
 	Description *string `json:"description,omitempty"`
+
+	// Tags is a list of tags to associate with the project.
+	Tags []string `json:"tags,omitempty"`
```
Comment (jtopjian) [inline openstack/identity/v3/projects/requests.go]: How can someone remove _all_ tags from a project? Should this be `*[]string`?

File: openstack/identity/v3/projects/testing/fixtures.go [hunk 2]
```diff
@@ -21,7 +21,8 @@ const ListOutput = `
       "enabled": true,
       "id": "1234",
       "name": "Red Team",
-      "parent_id": null
+      "parent_id": null,
+	  "tags": ["Red", "Team"]
```
Comment (jtopjian) [inline openstack/identity/v3/projects/testing/fixtures.go]: nit: spacing

File: openstack/identity/v3/projects/testing/fixtures.go [hunk 3]
```diff
@@ -30,7 +31,7 @@ const ListOutput = `
       "enabled": true,
       "id": "9876",
       "name": "Blue Team",
-      "parent_id": null
+	  "parent_id": null
```
Comment (jtopjian) [inline openstack/identity/v3/projects/testing/fixtures.go]: nit: spacing

File: openstack/identity/v3/projects/requests.go (hunk 1 above)
Comment (randomswdev) [reply openstack/identity/v3/projects/requests.go]: Just reviewed other resources; they tend to use pointers when updating tags. So I aligned project to those other resources and implemented the change in a new commit.

File: openstack/identity/v3/projects/testing/fixtures.go (hunk 3 above)
Comment (randomswdev) [reply openstack/identity/v3/projects/testing/fixtures.go]: Done in a new commit.

**Comment by randomswdev** [comment]
> In addition, we'll need to have the `doc.go` file updated to show examples of how to use these additions. And if possible, can you add some acceptance tests?

In my last two commits I added an acceptance test that executes few tag related operations and I updated doc.go

File: openstack/identity/v3/projects/requests.go [hunk 4]
```diff
@@ -32,6 +32,18 @@ type ListOpts struct {
...
+	Tags *[]string `q:"tags"`
+
+	// TagsAny filters on specific project tags. At least one of the tags must be present for the project.
+	TagsAny *[]string `q:"tags-any"`
+
+	// NotTags filters on specific project tags. All tags must be absent for the project.
+	NotTags *[]string `q:"not-tags"`
+
+	// NotTagsAny filters on specific project tags. At least one of the tags must be absent for the project.
+	NotTagsAny *[]string `q:"not-tags-any"`
```
Comment (jtopjian) [inline openstack/identity/v3/projects/requests.go]: Why are these pointers?

Comment (randomswdev) [reply openstack/identity/v3/projects/requests.go]: I reviewed the ListOpt of the other resources and aligned project to them: so I removed the * and changed the array to a plain string. Sorry for the error.

Comment (jtopjian) [reply openstack/identity/v3/projects/requests.go]: No problem at all!

**Comment by jtopjian** [comment]
@randomswdev Thanks!

I had to make the following changes for the acceptance tests to work:

```diff
--- a/acceptance/openstack/identity/v3/projects_test.go
+++ b/acceptance/openstack/identity/v3/projects_test.go
@@ -224,7 +224,7 @@ func TestProjectsTags(t *testing.T) {
        for _, project := range allProjects {
                tools.PrintResource(t, project)

-               if project.Name == "admin" {
+               if project.Name == projectMain.Name {
                        found = true
                }
        }
@@ -259,7 +259,7 @@ func TestProjectsTags(t *testing.T) {
        for _, project := range allProjects {
                tools.PrintResource(t, project)

-               if project.Name == "admin" {
+               if project.Name == projectMain.Name {
                        found = true
                }
        }
@@ -277,7 +277,16 @@ func TestProjectsTags(t *testing.T) {
        allProjects, err = projects.ExtractProjects(allPages)
        th.AssertNoErr(t, err)

-       th.AssertEquals(t, len(allProjects), 0)
+       found = false
+       for _, project := range allProjects {
+               tools.PrintResource(t, project)
+
+               if project.Name == projectMain.Name {
+                       found = true
+               }
+       }
+
+       th.AssertEquals(t, found, false)

        // Search matching not all tags
        listOpts = projects.ListOpts{
@@ -325,4 +334,15 @@ func TestProjectsTags(t *testing.T) {
        tools.PrintResource(t, updatedProject)
        th.AssertEquals(t, len(updatedProject.Tags), 1)
        th.AssertEquals(t, updatedProject.Tags[0], "Tag1")
+
+       // Remove all Tags
+       updateOpts = projects.UpdateOpts{
+               Tags: &[]string{},
+       }
+
+       updatedProject, err = projects.Update(client, projectMain.ID, updateOpts).Extract()
+       th.AssertNoErr(t, err)
+
+       tools.PrintResource(t, updatedProject)
+       th.AssertEquals(t, len(updatedProject.Tags), 0)
 }
```

The last part is an addition to make sure that removing all tags is successful.

Our OpenLab tests have also been fixed, so you'll be able to see the test output in the OpenLab logs.

Let me know if you have any questions.

**Comment by randomswdev** [comment]
> Our OpenLab tests have also been fixed, so you'll be able to see the test output in the OpenLab logs.
> 
> Let me know if you have any questions.

Impemented the changes you suggested. I'm waiting for the tests to run and complete successfully.

**Pull request merged**